### PR TITLE
proxy timings

### PR DESCRIPTION
**Pull request by i110**
Add a logging feature to log access timings to upstream servers

- `%{proxy-idle-time}x`
- `%{proxy-connect-time}x`
- `%{proxy-request-header-time}x`
- `%{proxy-request-body-time}x`
- `%{proxy-request-total-time}x`
- `%{proxy-first-byte-time}x`
- `%{proxy-response-time}x`
- `%{proxy-total-time}x`

TODO:
- [x] more granularity? (i.e. upstream-request-header-time, etc) It might needs some interface changes of http1client..
- [x] server-timings header
- [x] tests

### CODE REVIEW

File: include/h2o.h [hunk 1]
```diff
@@ -1085,6 +1087,13 @@ struct st_h2o_req_t {
      */
     H2O_VECTOR(h2o_req_error_log_t) error_logs;
 
+    /**
+     * handler specific log data
+     */
+    struct {
+        h2o_proxy_log_data_t *proxy;
+    } handler_log_data;
```
Comment (kazuho) [inline include/h2o.h]: Two questions here:
* does `proxy` need to be a pointer?
* do we need to wrap `h2o_proxy_log_data_t` with a struct named `handler_log_data`? I'd assume that the timing information we will be logging directly in h2o to be i) when the information is sent to the client and ii) when the information is generated by the generator. If my assumption is true, then we will have nothing other than `proxy`.

File: include/h2o/http1client.h [hunk 2]
```diff
@@ -57,6 +57,16 @@ typedef struct st_h2o_http1client_ctx_t {
...
 
+
+typedef struct st_h2o_http1client_timings_t {
+    uint64_t start_at;
+    uint64_t request_begin_at;
+    uint64_t request_body_begin_at;
+    uint64_t request_end_at;
+    uint64_t response_start_at;
+    uint64_t response_end_at;
+} h2o_http1client_timings_t;
```
Comment (kazuho) [inline include/h2o/http1client.h]: LGTM.

File: lib/core/logconf.c [hunk 3]
```diff
@@ -262,6 +264,14 @@ h2o_logconf_t *h2o_logconf_compile(const char *fmt, int escape, char *errbuf)
...
                     MAP_EXT_TO_PROTO("ssl.cipher-bits", ssl.cipher_bits);
                     MAP_EXT_TO_PROTO("ssl.session-id", ssl.session_id);
+
+                    h2o_log_handler_callback_t cb;
+                    if ((cb = h2o_proxy_get_logconf_callback(h2o_iovec_init(pt, quote_end - pt))) != NULL) {
+                        NEW_ELEMENT(ELEMENT_TYPE_HANDLER_SPECIFIC);
+                        LAST_ELEMENT()->data.handler_cb = cb;
+                        goto MAP_EXT_Found;
+                    }
+
```
Comment (kazuho) [inline lib/core/logconf.c]: Can't we simply add `ELEMENT_TYPE_PROXY_PROCESS_TIME` etc. due to the same reason stated above?

**Comment by i110** [comment]
Thank you for your comment. As you said I added a little abstraction because
 
1. I wanted to encapsulate actual logging logics in `proxy.c`. I anticipated that additional demands of logging like `upstream-addr` or `upstream-status` in nginx will be raised soonly. If that happen `logconf.c` will get too completed to maintain unless this encapsulating I think
2. other handler might want to add logging values in the future. `%{mruby.foo}x` or `%{file.bar}x` may be useful sometime

**Comment by kazuho** [comment]
Thank you for the answer.

The problem as I see is that the abstraction is incomplete. For example, existence of handler-specific type (i.e. `h2o_req_t::handler_log_data.proxy`) means that the feature is not handler-agnostic, while some of the functions (e.g. logger callbacks) are tried to be abstracted.

I also wonder it would be the right approach to let the handler define a dedicated callback for each log specifier would be the right approach for a _genenic mechanism_. For example, you might want to generate an arbitrary-named property in the mruby handler. But you cannot dynamically create a callback that binds to the named property.

In other words, I am backwards against trying to design a framework at the moment when there is no clear need for one.

**Comment by i110** [comment]
> the feature is not handler-agnostic, while some of the functions (e.g. logger callbacks) are tried to be abstracted.

I agree, it's sort of like result of compromise. I don't have objection to removing `handler_log_data` struct

> For example, you might want to generate an arbitrary-named property in the mruby handler. But you cannot dynamically create a callback that binds to the named property.

In such a case, for example, mruby handler can catch `mruby-` prefix and run the callback written in mruby. Only thing I wanted to do is encapsulating the logics in each handlers, not developing framework.

File: include/h2o.h [hunk 4]
```diff
@@ -1085,6 +1085,11 @@ struct st_h2o_req_t {
      */
     H2O_VECTOR(h2o_req_error_log_t) error_logs;
 
+    /**
+     * proxy log data
+     */
+    h2o_http1client_timings_t *proxy_timings;
```
Comment (kazuho) [inline include/h2o.h]: Thank you for the changes.

How about adding `h2o_http1client_timings_t` (not a pointer) as a member of `h2o_req_t::timestamps`?

By doing so, I believe that we can drastically reduce the changes. For example, there would be no need for new functions like `h2o_proxy_time_compute_...`. All we need to do in core is to support new variables in logconf.c.

**Comment by i110** [comment]
@kazuho Thank you for your comment. I addressed the issues.

File: lib/core/util.c [hunk 5]
```diff
@@ -828,9 +828,12 @@ void h2o_add_server_timing_header(h2o_req_t *req)
...
-    ELEMENT_LONGEST_STR(connect)                                                                                                   \
-    DELIMITER ELEMENT_LONGEST_STR(header) DELIMITER ELEMENT_LONGEST_STR(body) DELIMITER ELEMENT_LONGEST_STR(request_total)         \
-        DELIMITER ELEMENT_LONGEST_STR(process) DELIMITER ELEMENT_LONGEST_STR(response)
+    ELEMENT_LONGEST_STR("connect")                                                                                                 \
+    DELIMITER ELEMENT_LONGEST_STR("header") DELIMITER ELEMENT_LONGEST_STR("body") DELIMITER ELEMENT_LONGEST_STR("request_total")   \
+        DELIMITER ELEMENT_LONGEST_STR("process") DELIMITER ELEMENT_LONGEST_STR("proxy-idle")                                       \
+            DELIMITER ELEMENT_LONGEST_STR("proxy-connect") DELIMITER ELEMENT_LONGEST_STR("proxy-request-header")                   \
+                DELIMITER ELEMENT_LONGEST_STR("proxy-request-body") DELIMITER ELEMENT_LONGEST_STR("proxy-request-total")           \
+                    DELIMITER ELEMENT_LONGEST_STR("proxy-first-byte")
```
Comment (kazuho) [inline lib/core/util.c]: Do these field names match those that are actually being emitted?

Comment (i110) [reply lib/core/util.c]: oops, thank you for finding this. Fixed at https://github.com/h2o/h2o/pull/1717/commits/06b4a9bbf24aa8310dd27db8f7b12a226952855e
And also I found similar bug exists in h2o_build_server_timing_trailer, then I fixed it at https://github.com/h2o/h2o/pull/1717/commits/ce33863ffeb361bae528af629251d829e95cec65

**Comment by kazuho** [comment]
Thank you for the changes. Merged to master.
